auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/external/DeltaResetPasswordController.kt

```diff
@@ -0,0 +1,228 @@
+
+    private suspend fun resetPasswordExpiredPost(call: ApplicationCall) {
+        val formParameters = call.receiveParameters()
+        val userCN = formParameters["userCN"].orEmpty()
+        val token = formParameters["token"].orEmpty()
+        val tokenResult = resetPasswordTokenService.consumeToken(token, userCN)
+        if (tokenResult is PasswordTokenService.ExpiredToken) {
+            logger.atInfo().addKeyValue("userCN", userCN).log("Sending new reset password link (after expiry)")
+            sendNewResetPasswordLink(userCN)
+            call.respondNewEmailSentPage(userCN.replace("!", "@"))
+        } else throw Exception("tokenResult was $tokenResult when trying to send a new reset password email")
+    }
+
+    private suspend fun resetPasswordPost(call: ApplicationCall) {
+        val userCN = call.request.queryParameters["userCN"].orEmpty()
+        val token = call.request.queryParameters["token"].orEmpty()
+
+        if (Strings.isNullOrEmpty(userCN)) throw ResetPasswordException(
+            "reset_password_no_user_cn",
+            "User CN not present on resetting password"
+        )
+
+        if (Strings.isNullOrEmpty(token)) throw ResetPasswordException(
+            "reset_password_no_token",
+            "Token not present on resetting password"
+        )
+
+        val (message, newPassword) = passwordChecker.checkPasswordForErrors(call, userCN)
+
+        if (message != null) return call.respondResetPasswordPage(message)
+        logger.atInfo().addKeyValue("userCN", userCN).log()
+        val tokenResult = resetPasswordTokenService.consumeToken(token, userCN)
+        call.respondToResult(tokenResult, newPassword)
+    }
+
+    private suspend fun sendNewResetPasswordLink(userCN: String) {
+        val user = userLookupService.lookupUserByCn(userCN)
+        logger.atInfo().addKeyValue("userCN", userCN).addKeyValue("emailAddress", user.email)
+            .log("Sending reset password link")
+        val token = resetPasswordTokenService.createToken(userCN)
+        emailService.sendTemplateEmail(
+            "reset-password",
+            EmailContacts(
+                user.email!!,
+                user.fullName,
+                emailConfig.fromEmailAddress,
+                emailConfig.fromEmailName,
+                emailConfig.replyToEmailAddress,
+                emailConfig.replyToEmailName,
+            ),
+            "DLUHC DELTA - Reset Your Password",
+            mapOf(
+                "deltaUrl" to deltaConfig.deltaWebsiteUrl,
+                "userFirstName" to user.firstName,
+                "resetPasswordUrl" to getResetPasswordURL(
+                    token,
+                    userCN,
+                    authServiceConfig.serviceUrl
+                )
+            )
+        )
+        logger.atInfo().addKeyValue("userCN", userCN).addKeyValue("emailAddress", user.email)
+            .log("Sent reset password link")
+    }
+
+    private suspend fun ApplicationCall.respondToResult(
```

Comment:
I don't like the name of this, it's doing more than responding to a result, it's actually doing the password reset.

Not sure this name is much better though

```suggestion
    private suspend fun ApplicationCall.handleResetPasswordTokenResult(
```